### PR TITLE
#206: Preserve environment markers from requirements.in

### PR DESCRIPTION
**Pull request by barrywhart**
* piptools/scripts/compile.py
  * When calling `OutputWriter.write()`, pass along the markers from `requirements.in`
* piptools/util.py
  * `format_requirement()` -- add support for optional markers
* piptools/writer.py
  * Add `marker` parameter to `OutputWriter.write()`, pass it to `piptools.util.format_requirement()`
* tests/test_writer.py
  * Test for the new behavior

### CODE REVIEW

**Comment by barrywhart** [comment]
For some reason, I can't view the Travis build log. The tests are passing locally for me. Can someone tell me what the Travis build error was?

**Comment by adamchainz** [comment]
From 2.7:

```
=================================== FAILURES ===================================
__________________ test_format_requirement_environment_marker __________________
from_line = <bound method type.from_line of <class 'pip.req.req_install.InstallRequirement'>>
writer = <piptools.writer.OutputWriter object at 0x7fcd08690850>
    def test_format_requirement_environment_marker(from_line, writer):
        "Primary packages should not get annotated."
        ireq = from_line("test ; python_version=='2.7' and platform_python_implementation == 'CPython'")
        reverse_dependencies = set()
    
>       assert (writer._format_requirement(ireq,
                                           reverse_dependencies,
                                           primary_packages=['test'],
                                           marker=ireq.markers) ==
                "test ; python_version=='2.7' and platform_python_implementation == 'CPython'")
E       assert 'test ; pytho... == "CPython"' == "test ; python... == 'CPython'"
E         - test ; python_version == "2.7" and platform_python_implementation == "CPython"
E         ?                      -  ^^   ^                                       ^       ^
E         + test ; python_version=='2.7' and platform_python_implementation == 'CPython'
E         ?                        ^   ^                                       ^       ^
tests/test_writer.py:65: AssertionError
===================== 1 failed, 52 passed in 1.02 seconds ======================

```

**Comment by adamchainz** [comment]
3.5 errored and seemed to produce no logs, if you push a new commit it'll restart the tests and it might succeed

**Comment by barrywhart** [comment]
Thanks, I'm seeing build results now. Fixed a test that was failing on Python 3.x and fixed a style checker complaint about lines that are too long.

**Comment by barrywhart** [comment]
Hmm, fixed a few problems, now Travis is no long showing build output. Travis is not being my friend.

**Comment by davidovich** [comment]
@barrywhart can you squash your commits ?

I think we can add this to the next feature release (1.9).

File: tests/test_writer.py [hunk 1]
```diff
@@ -55,3 +55,16 @@ def test_format_requirement_not_for_primary(from_line, writer):
...
+
+def test_format_requirement_environment_marker(from_line, writer):
+    "Primary packages should not get annotated."
+    ireq = from_line("test ; python_version == '2.7' and platform_python_implementation == 'CPython'")
+    reverse_dependencies = set()
+
+    result = writer._format_requirement(
+        ireq, reverse_dependencies, primary_packages=['test'],
+        marker=ireq.markers)
+    result = result.replace('\"', '\'')
```
Comment (adamchainz) [inline tests/test_writer.py]: I don't get why you're modifying the result before asserting on it, why not just assert on the original value of result?

Comment (barrywhart) [inline tests/test_writer.py]: I was seeing different behavior in Python 2.7 vs 3.5. Python 2.7 preserved quotes in the original format, while Python 3.5 was converting to double quotes. I assume this is a difference in the underlying pip library which parses the lines (including environment markers) from `requirements.in`.

**Comment by barrywhart** [comment]
Closing this PR. I had already done a "git pull" without "--rebase", which made it hard to squash commits. Here is a new PR with the commits squashed: https://github.com/jazzband/pip-tools/pull/460.

**Pull request closed without merging**